{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fbe6f142-f111-4d78-923b-025af552812d",
   "metadata": {},
   "source": [
    "eQTL mapping with tensorQTL. Try this if fastQTL does not work for you"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea8931c4-3900-4656-ac1b-7b0ec0282bc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! pip3 install tensorqtl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "8f59da3e-8571-4c2a-8841-aa858daffe9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from io import StringIO\n",
    "import subprocess\n",
    "import tensorqtl\n",
    "from tensorqtl import pgen, cis, trans, post"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2919f5c-91a8-467f-bada-f33f648dfef8",
   "metadata": {},
   "source": [
    "# 0. Test section (ignore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "63a29e1e-affc-4505-9c7d-eb6d9ef7e054",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HD0280_HA0023</th>\n",
       "      <th>BD2180_BD6180</th>\n",
       "      <th>BD2179_BD6179</th>\n",
       "      <th>BD2288_BD6288</th>\n",
       "      <th>HD0119_HD4119</th>\n",
       "      <th>HD0107_HD4107</th>\n",
       "      <th>LD0213_LA0013</th>\n",
       "      <th>BD3539_BA0523</th>\n",
       "      <th>LD0175_LD4175</th>\n",
       "      <th>BD2527_BD6527</th>\n",
       "      <th>...</th>\n",
       "      <th>HD0275_HA0018</th>\n",
       "      <th>BD1533_BD5533</th>\n",
       "      <th>BD2455_BD6455</th>\n",
       "      <th>LD0144_LD4144</th>\n",
       "      <th>BD2188_BD6188</th>\n",
       "      <th>BD2606_BD6606</th>\n",
       "      <th>LD0082_LD4082</th>\n",
       "      <th>BD2833_BD6833</th>\n",
       "      <th>BD3346_BA0352</th>\n",
       "      <th>HD0145_HD4145</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr7:12795:C:T</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr7:20992:C:G</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr7:21264:C:T</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 1607 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                HD0280_HA0023  BD2180_BD6180  BD2179_BD6179  BD2288_BD6288  \\\n",
       "chr7:12795:C:T              0              0              0              0   \n",
       "chr7:20992:C:G              0              0              0              0   \n",
       "chr7:21264:C:T              0              0              0              0   \n",
       "\n",
       "                HD0119_HD4119  HD0107_HD4107  LD0213_LA0013  BD3539_BA0523  \\\n",
       "chr7:12795:C:T              0              0              0              0   \n",
       "chr7:20992:C:G              0              0              0              0   \n",
       "chr7:21264:C:T              0              0              0              0   \n",
       "\n",
       "                LD0175_LD4175  BD2527_BD6527  ...  HD0275_HA0018  \\\n",
       "chr7:12795:C:T              0              0  ...              0   \n",
       "chr7:20992:C:G              0              0  ...              0   \n",
       "chr7:21264:C:T              0              0  ...              0   \n",
       "\n",
       "                BD1533_BD5533  BD2455_BD6455  LD0144_LD4144  BD2188_BD6188  \\\n",
       "chr7:12795:C:T              0              0              0              0   \n",
       "chr7:20992:C:G              0              0              0              0   \n",
       "chr7:21264:C:T              0              0              0              0   \n",
       "\n",
       "                BD2606_BD6606  LD0082_LD4082  BD2833_BD6833  BD3346_BA0352  \\\n",
       "chr7:12795:C:T              0              0              0              0   \n",
       "chr7:20992:C:G              0              0              0              0   \n",
       "chr7:21264:C:T              0              0              0              0   \n",
       "\n",
       "                HD0145_HD4145  \n",
       "chr7:12795:C:T              0  \n",
       "chr7:20992:C:G              0  \n",
       "chr7:21264:C:T              0  \n",
       "\n",
       "[3 rows x 1607 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr7:12795:C:T</th>\n",
       "      <td>7</td>\n",
       "      <td>12795</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr7:20992:C:G</th>\n",
       "      <td>7</td>\n",
       "      <td>20992</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr7:21264:C:T</th>\n",
       "      <td>7</td>\n",
       "      <td>21264</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               chrom    pos  index\n",
       "id                                \n",
       "chr7:12795:C:T     7  12795      0\n",
       "chr7:20992:C:G     7  20992      1\n",
       "chr7:21264:C:T     7  21264      2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Test load plink pgen data\n",
    "# PLINK reader for genotypes\n",
    "plink_prefix_path = '/data100t1/home/wanying/CCHC/eQTL_gtex_pipeline/code/example_data/cchc_chr8_example'\n",
    "pgr = pgen.PgenReader(plink_prefix_path)\n",
    "genotype_df = pgr.load_genotypes()\n",
    "display(genotype_df.head(3))\n",
    "\n",
    "variant_df = pgr.variant_df\n",
    "display(variant_df.head(3))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f120bceb-151f-4c70-be30-7b2874065daa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>1600</th>\n",
       "      <th>1601</th>\n",
       "      <th>1602</th>\n",
       "      <th>1603</th>\n",
       "      <th>1604</th>\n",
       "      <th>1605</th>\n",
       "      <th>1606</th>\n",
       "      <th>1607</th>\n",
       "      <th>1608</th>\n",
       "      <th>1609</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7:12795:C:T</td>\n",
       "      <td>chr7</td>\n",
       "      <td>12795</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>...</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7:20992:C:G</td>\n",
       "      <td>chr7</td>\n",
       "      <td>20992</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>...</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0|0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 1610 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0     1      2    3    4    5    6    7    8    9     ... 1600  \\\n",
       "0  chr7:12795:C:T  chr7  12795  0|0  0|0  0|0  0|0  0|0  0|0  0|0  ...  0|0   \n",
       "1  chr7:20992:C:G  chr7  20992  0|0  0|0  0|0  0|0  0|0  0|0  0|0  ...  0|0   \n",
       "\n",
       "  1601 1602 1603 1604 1605 1606 1607 1608 1609  \n",
       "0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
       "1  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
       "\n",
       "[2 rows x 1610 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get genotypes by bcftools (for fun)\n",
    "vcf_fn = '/data100t1/home/wanying/CCHC/eQTL_gtex_pipeline/code/example_data/cchc_chr8_example.vcf.gz'\n",
    "cmd = f\"bcftools query -f %ID,%CHROM,%POS[,%GT]\\\\n {vcf_fn}\"\n",
    "result = subprocess.run(cmd.split(), capture_output=True, text=True, check=True).stdout\n",
    "\n",
    "gp_data = StringIO(result)\n",
    "df = pd.read_csv(gp_data, header=None)\n",
    "df.head(2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef98acb9-dac7-4617-ae60-d00860810332",
   "metadata": {},
   "source": [
    "# 1. Run tensorQTL\n",
    "Test run on chr21"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c77e9d0f-8ab0-47ae-aedf-b0055d59522d",
   "metadata": {},
   "source": [
    "## 1.1 Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "b6e2492c-f89c-46c5-8e2c-26ce9477ad0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define paths to data\n",
    "plink_prefix_path = '/data100t1/home/wanying/CCHC/CHARGE_GWAS/GEM_genotype_files/CCHC_chr10'\n",
    "expression_bed = '/data100t1/home/wanying/CCHC/eQTL_gtex_pipeline/input/chr_col_fixed_tmm/CCHC_batch1_2_3_4_TMM.chr10.chr_fixed.bed.gz'\n",
    "covariates_file = '/data100t1/home/wanying/CCHC/eQTL_gtex_pipeline/supporting_files/combined_covariates_peers_from_different_runs/redo_combined_covar_sex_age_pc1-5.60_peers.combined_covariates.txt'\n",
    "prefix = 'test_run_chr10'\n",
    "\n",
    "# load phenotypes and covariates\n",
    "phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expression_bed)\n",
    "covariates_df = pd.read_csv(covariates_file, sep='\\t', index_col=0).T\n",
    "\n",
    "# PLINK reader for genotypes\n",
    "pgr = pgen.PgenReader(plink_prefix_path)\n",
    "genotype_df = pgr.load_genotypes()\n",
    "variant_df = pgr.variant_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc762130-9cf7-45ab-9990-61dddd64f5bc",
   "metadata": {},
   "source": [
    "## 1.2 Nominal run in cis region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "dce4bbff-93f6-44c9-b22c-b57fdb5aba56",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 1288 samples\n",
      "  * 1002 phenotypes\n",
      "  * 67 covariates\n",
      "  * 3315380 variants\n",
      "  * cis-window: Â±1,000,000\n",
      "  * checking phenotypes: 1002/1002\n",
      "  * Computing associations\n",
      "    Mapping chromosome 10\n",
      "    processing phenotype 1002/1002\n",
      "    time elapsed: 19.11 min\n",
      "    * writing output\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# map all cis-associations (results for each chromosome are written to file)\n",
    "# all genes\n",
    "cis_df = cis.map_nominal(genotype_df, variant_df,\n",
    "                         phenotype_df, phenotype_pos_df,\n",
    "                         prefix, covariates_df=covariates_df,\n",
    "                         output_dir='/data100t1/home/wanying/CCHC/eQTL_gtex_pipeline/output/tensorQTL_runs/')\n",
    "\n",
    "# # Only run genes on one chromomsome\n",
    "# cis.map_nominal(genotype_df, variant_df,\n",
    "#                 phenotype_df.loc[phenotype_pos_df['chr'] == 'chr18'],\n",
    "#                 phenotype_pos_df.loc[phenotype_pos_df['chr'] == 'chr18'],\n",
    "#                 prefix, covariates_df=covariates_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "54c7fae3-d983-460c-8557-f5a2bdd1af8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>start_distance</th>\n",
       "      <th>af</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000261456.5</td>\n",
       "      <td>chr10:10537:C:A</td>\n",
       "      <td>-63626</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000261456.5</td>\n",
       "      <td>chr10:10537:C:T</td>\n",
       "      <td>-63626</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000261456.5</td>\n",
       "      <td>chr10:10550:G:T</td>\n",
       "      <td>-63613</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000261456.5</td>\n",
       "      <td>chr10:10552:TCAG:T</td>\n",
       "      <td>-63611</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000261456.5</td>\n",
       "      <td>chr10:10554:A:G</td>\n",
       "      <td>-63609</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        phenotype_id          variant_id  start_distance   af  ma_samples  \\\n",
       "0  ENSG00000261456.5     chr10:10537:C:A          -63626  0.0           0   \n",
       "1  ENSG00000261456.5     chr10:10537:C:T          -63626  0.0           0   \n",
       "2  ENSG00000261456.5     chr10:10550:G:T          -63613  0.0           0   \n",
       "3  ENSG00000261456.5  chr10:10552:TCAG:T          -63611  0.0           0   \n",
       "4  ENSG00000261456.5     chr10:10554:A:G          -63609  0.0           0   \n",
       "\n",
       "   ma_count  pval_nominal  slope  slope_se  \n",
       "0         0           NaN    NaN       NaN  \n",
       "1         0           NaN    NaN       NaN  \n",
       "2         0           NaN    NaN       NaN  \n",
       "3         0           NaN    NaN       NaN  \n",
       "4         0           NaN    NaN       NaN  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load results\n",
    "output_dir = '/data100t1/home/wanying/CCHC/eQTL_gtex_pipeline/output/tensorQTL_runs'\n",
    "pairs_df = pd.read_parquet(f'{output_dir}/{prefix}.cis_qtl_pairs.10.parquet')\n",
    "pairs_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "969242be-bbf6-4904-b8f9-7e7b29bad3a4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
